src/Shared/FileMatcher.cs

```diff
@@ -705,128 +826,183 @@ private static void GetFilesRecursive
                         return;
                     }
                 }
             }
 
-            RecursiveStepResult nextStep = GetFilesRecursiveStep(
-                recursionState,
-                projectDirectory,
-                stripProjectDirectory,
-                getFileSystemEntries);
+            RecursiveStepResult nextStep = GetFilesRecursiveStep(recursionState);
 
-            RecursiveStepResult[] excludeNextSteps = null;
-            if (searchesToExclude != null)
+            IList<string> files = null;
+            foreach (string file in GetFilesForStep(nextStep, recursionState, projectDirectory,
+                stripProjectDirectory, getFileSystemEntries))
             {
-                excludeNextSteps = new RecursiveStepResult[searchesToExclude.Count];
-                for (int i = 0; i < searchesToExclude.Count; i++)
+                if (excludeNextSteps == null)
                 {
-                    excludeNextSteps[i] = GetFilesRecursiveStep(
-                        searchesToExclude[i],
-                        projectDirectory,
-                        stripProjectDirectory,
-                        getFileSystemEntries);
+                    files = files ?? new List<string>();
+                    files.Add(file);
+                    continue;
                 }
-            }
-
-            if (nextStep.Files != null)
-            {
-                HashSet<string> filesToExclude = null;
-                if (excludeNextSteps != null)
+                var exclude = false;
+                for (int i = 0; i < excludeNextSteps.Length; i++)
                 {
-                    filesToExclude = new HashSet<string>();
-                    foreach (var excludeStep in excludeNextSteps)
+                    RecursiveStepResult excludeNextStep = excludeNextSteps[i];
+                    if (excludeNextStep.ConsiderFiles && MatchFileRecursionStep(searchesToExclude[i], file))
                     {
-                        if (excludeStep.Files != null)
-                        {
-                            foreach (var file in excludeStep.Files)
-                            {
-                                filesToExclude.Add(file);
-                            }
-                        }
+                        exclude = true;
+                        break;
                     }
                 }
-
-                foreach (var file in nextStep.Files)
+                if (!exclude)
                 {
-                    if (filesToExclude == null || !filesToExclude.Contains(file))
-                    {
-                        listOfFiles.Add(file);
-                    }
+                    files = files ?? new List<string>();
+                    files.Add(file);
                 }
             }
+            // Add all matched files at once to reduce thread contention
+            if (files != null && files.Any())
+            {
+                listOfFiles.PushRange(files.ToArray());
```

Comment:
Instead of copying the list to an array, and then create a link list node for each of them (inside the stack code), I think it can be Stack<List<string>>(), so you just push the non-empty list there.  In the code consuming it, it will be:
stack.SelectMany(list => list).xxxx..., so we can prevent the extra allocation, in case the list is big.

Reply:
[Done](https://github.com/Microsoft/msbuild/pull/2572/commits/b5e4a47c725f3dbbec16b3fda6af9fd69175ea58)